File: Algorithms_python/62_Unique_Paths.py
class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        dp = [[1] * n for _ in range(m)]
        for i in range(1, m):
            for j in range(1, n):
                dp[i][j] = dp[i - 1][j] + dp[i][j - 1]
        return dp[-1][-1]
'''
动态规划题，想法是：
假设 dp[i][j] 表示到达网格 (i,j) 的路径数，则 dp[i][j] = dp[i - 1][j] + dp[i][j - 1]。
可以画个表格模拟一下，每一个能到达的路径数，等于他的上左两格相加
先把第所有格子全设成1
'''

File: Algorithms_C#/62_Unique_Paths.cs
public class Solution {
    public int UniquePaths(int m, int n) {
        int[][] f = new int[m][];
        for (int i = 0; i < m; i++) 
        {
            f[i] = new int[n];
        }
        f[0][0] = 1;
        for (int i  = 0; i < m; ++i){
            for (int j = 0; j < n; ++j){
                if (i > 0){
                    f[i][j] += f[i-1][j];
                }
                if (j > 0){
                    f[i][j] += f[i][j-1];
                }
            }
        }
        return f[m-1][n-1];
    }
}